### ISSUE TITLE
Multi-Language Language Server

### ISSUE THREAD

**Issue by timfel** [Microsoft/language-server-protocol#636]
@numberpi is creating a language server that exposes the GraalVM languages (https://github.com/numberpi/graal/tree/lsp-pr). This server naturally supports multiple languages. It's not clear at this point how this is supposed to work with the LSP protocol, given that things like completion trigger characters have to be sent to the client without any connection to the language.

Is the recommended approach then to spawn multiple servers, one for each language? Or is an extension of the LSP protocol possible that would allow us to support multiple languages with different language-specific trigger characters?

I guess this is somewhat related to https://github.com/Microsoft/language-server-protocol/issues/505, although in our case the same server could be used to support multiple languages mixed in the same file.

**Comment by mickaelistria** [comment]
That's an excellent use-case. As far as I know, there is no real recommendation for this story; and it's very likely that you're going to be a good candidate to build a recommendation according to the experience you'll collect with this use case ;)
One thing is that one connection between an IDE and a LS seems to be really locked down to one language/feature set by the spec; as you noticed. So it doesn't seem like achieving multi-language is something that can be done by a single LS connection.
Have you considered the idea of language-server "proxies" or "facades"? One possibility would be to have a single language server (the Graal one) acting as a service and to have several language-specific language servers the IDE would connect to. So for .java file, it would invoke the Java Graal language server which would take care of providing Java-specific content, and this Java Graal Language Server would basically proxy and filter messages to/from the common Graal server; same thing for JS or any language that can use Graal.
Does it sound like a possible way forward?

**Comment by chumer** [comment]
> Does it sound like a possible way forward?

Yes. We thought about this as well as one of the ways forward. Thanks for the response. Two more questions:

What would the consequences of this be for debugging integration? For debugging we need one session for all languages. Or is this always an entirely separate protocol anyway?

How could we deal with multiple languages in the same file in this scenario? For example, with GraalVM we could tell you the language of a string literal when its eval'ed. For example in JS you can write:
`Polyglot.eval("python", "21 + 21")` and we could tell you the section of the program that is actually python code (only if the string is not dynamic of course). And we could auto-complete in there as well.

Off Topic: we had a great time so far connecting Graal languages using LSP to Netbeans and VS code. We plan to make it a standard feature in GraalVM. Thanks for all the work you do.

**Comment by mickaelistria** [comment]
> What would the consequences of this be for debugging integration?

Do you mean you (LS/IDE developer) debugging your LSP-based integration, or user debugging over LSP?
If the former, then it's all a black-box, the protocol doesn't enforce anything; it would most likely be that you'd look a messages between IDE and proxy, and between proxy and central Graal LS and check how those are mapped correctly for the proxy part, and returning right values for the pure language part.
If the later, then running and debugging target applications is another story that's not part of the LSP, but which is part of the Debug Adapter Protocol, another protocol, another GitHub repo, another issue ;)

> How could we deal with multiple languages in the same file in this scenario?

The VSCode HTML Language Server have support for CSS and JS inside HTML files. It happens to work, I'm not sure how, but I doubt it actually uses generic LSP composition approach and guess it actually just embeds some CSS and JS logic inside the HTML server (instead of delegating to other LS).
For Graal, since the idea is that the LS is multi-platform, I think it's just a matter of having your GraalLS recognizing the language in the String a reporting the right content. No need to delegate from a Java proxy to a Python proxy if in the end, they're going to use the same backend.

**Comment by chumer** [comment]
> Do you mean you (LS/IDE developer) debugging your LSP-based integration, or user debugging over LSP?

I mean the latter. 

> If the later, then running and debugging target applications is another story that's not part of the LSP, but which is part of the Debug Adapter Protocol, another protocol, another GitHub repo, another issue ;)

Got it, thanks! Currently, we use the chrome inspector debugging protocol bridge. But we have some plans to support the Debug Adapter Protocol directly.

> For Graal, since the idea is that the LS is multi-platform, I think it's just a matter of having your GraalLS recognizing the language in the String a reporting the right content. No need to delegate from a Java proxy to a Python proxy if in the end, they're going to use the same backend.

But then, back to Tims original question, how do we change the autocompletion trigger for that source section?

**Comment by mickaelistria** [comment]
> But then, back to Tims original question, how do we change the autocompletion trigger for that source section?

I just suggest you put about every char for any language as the completion trigger and ignore requests that seem alien to the current scope.
In many IDEs nowadays, users expect completion all the time, so the completion trigger chars can be a very greedy set and users would like it anyway ;)

But this is indeed a limitation of the protocol.

Do you have other examples of what can be affected other than completion trigger chars? It'd be good to have a list of "context specific" values that the protocol currently assume are static final.

**Comment by numberpi** [comment]
> I just suggest you put about every char for any language as the completion trigger and ignore requests that seem alien to the current scope.

How to ignore a request? We have to respond something. We already try this approach in our current implementation, so that when a code completion is triggered at client-side and it is an unknown trigger char for the current language, we tried to respond an empty list or the previous list of completion items. But (at least in Visual Studio Code) this interrupts other running actions at client-side (for example a former code completion).
A good example is mixing the languages Python and R. R allows dots in variable identifiers. For Python, it is a property completion trigger character. When writing R code, as soon as a dot is typed during a running completion, a new completion request is sent, cancelling the old one. Even if the new one might contain the same list of completion items as before, the already typed characters (including the trigger char) are then not taken to filter the list of suggestions anymore.

> Do you have other examples of what can be affected other than completion trigger chars? It'd be good to have a list of "context specific" values that the protocol currently assume are static final.

Basically the list can be derived from all things, which have to be defined statically up-front in "ServerCapabilities". But in my opinion, the ones which have an impact on the current Graal-LSP implementation are only completion trigger characters and signatureHelp trigger characters. Other things like "ExecuteCommandOptions" are also defined statically. But as they are only included in responses of other requests, this should not interfere with our multi-language LS.

**Comment by dbaeumer** [comment]
Registering for different documentsis actually supported in the LSP however it requires that the client opts into the dynamic registration support that got added in 3.0 version of the spec. If a client supports that you can register code complete (and all other code smarts) for n different document selectors each with different trigger characters.

> How could we deal with multiple languages in the same file in this scenario?

There is no generic composition mechanism in the LSP. However you can register two completion item providers for the same document using different trigger character sets. In VS Code only the one will be called that actually defines the trigger character. However I am not sure if this will be the case in all clients and that we need to better document this in the spec.

**Comment by kdvolder** [comment]
> How could we deal with multiple languages in the same file in this scenario?

Besides @dbaeumer suggestion. It is also possible for a language server to determine exactly what to do and 'dispatch' the completion request internally based on whatever logic it sees fit.

So if your language server knows what kind of content the document can have (and really, it should because how else can it do its job properly)...  The it should be able to partition the document, decide what 'language' is being used at the cursor location it is asked to provide completions for. Then based on that knowledge it can then provide completions specific to that context.
